lib/acl.js

```diff
@@ -43,16 +42,35 @@ Acl.prototype.query = function(role, resource, action, done) {
     : getParentRoles.call(this, extractedRole);
 
   resources = getParentResources.call(this, resource);
-
-  for (var k = this.permissions.length - 1; k >= 0; k--)
-    for (var i in roles)
-      for (var j in resources)
-        if (this.permissions[k].match(roles[i] || null, resources[j] || null, action || null)) {
-          matches.push(this.permissions[k]);
+  
+  
+  // Note: 
+  // Instead of creating the permission list 
+  // and then evaluating it we evaluate right away in LIFO order 
+  // The logic is the same as original however we avoid creating extra objects 
+  // and iterating second time over the matched permissions 
+  // this also avoids setTimeout from PermissionList 
+  const perm = this.permissions;
+  let k = perm.length - 1
+  const next = () => {
+    // advance to next matching rule (LIFO)
+    for (; k >= 0; k--) {
+      for (let i = 0; i < roles.length; i++) {
+        for (let j = 0; j < resources.length; j++) {
```

Comment:
Can you name the variables so it's easier to read?

```suggestion
  let permIndex = perm.length - 1
  const next = () => {
    // advance to next matching rule (LIFO)
    for (; permIndex >= 0; permIndex--) {
      for (let roleIndex = 0; roleIndex < roles.length; i++) {
        for (let resourceIndex = 0; resourceIndex < resources.length; j++) {
```

Reply:
done